RFC: Tofu Version Compatibility

cam72cam opened the pull request:
In this RFC, we investigate current problems around software versioning and propose some potential solutions.

This is contingent on the [.tofu extension RFC](https://github.com/opentofu/opentofu/pull/1699) [merged]

Resolves #1708

## Target Release

1.10.0

## Checklist

<!-- Please check of ALL items in this list for all PRs: -->

- [x] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [x] I have not used an AI coding assistant to create this PR.
- [x] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.

cam72cam committed:
Add RFC Tofu Version Compatibility

Signed-off-by: Christian Mesh <christianmesh1@gmail.com>

cam72cam commented on rfc/20240610-Tofu-Version-Compatibility.md:
````diff
@@ -0,0 +1,109 @@
+```shell
+$ tofu init
+...
+Warning: Using v1.7.x in 'terraform -> required_version' as equivalent to current tofu version 1.6.5!
````
It's tricky though, we will need to communicate why the "required_version" constraint was not met in some scenarios.

We could split it into two messages, a warning for specifying `terraform -> required_version` and one that's more detailed when the constraint is not met.

ollevche replied:
I think we need to reconsider the version mapping approach. It will become obsolete as soon as Terraform releases an incompatible feature.

After that, OpenTofu will still depend on module authors to update their configuration with alternative `.tofu` implementations.

Also, required version could be incompatible, but an actual configuration is fine to run with OpenTofu and Terraform (if it doesn't use incompatible features).

ImIOImI commented:
Although I don't think this is critical now because very few popular module authors require very recent versions of Terraform, as the two projects start to diverge it will become increasingly more important. Its better to have a plan in place and be able to revise it while its less important than implement a new system when it is critical.

cam72cam committed:
Update to reflect most recent conversation

Signed-off-by: Christian Mesh <christianmesh1@gmail.com>

cam72cam commented on rfc/20240610-Tofu-Version-Compatibility.md:
```diff
@@ -0,0 +1,106 @@
+
+## Proposed Solution
+
+As we introduce support for `.tofu` files, we open up the ability to add new language features without worrying as much about compatibility in shared modules. If we can differentiate between "OpenTofu Version" and "Terraform Version", the above scenarios become much easier.
```
I will once I can rebase this PR once the .tofu RFC is merged.

ImIOImI commented on rfc/20240610-Tofu-Version-Compatibility.md:
```diff
@@ -0,0 +1,106 @@
+Implement an option to disable required_version checking.
+
+Try to guess the equivalent OpenTofu version for a specified Terraform version.
+
```
Building a GitHub action was proposed as a possible alternative action as well. I particularly like this no matter what solution we land on because I'm assuming I can use it to guard against unexpected incompatibilities in modules I include in my code as well when I update versions.

ollevche replied:
We also discussed some sort of compatibility table between versions and modules based on real user feedback. I know it's tricky, but still, it has some practical benefits. I would add those to the open questions block

cam72cam replied:
Noted in b3b094db80

skyzyx commented on rfc/20240610-Tofu-Version-Compatibility.md:
```diff
@@ -0,0 +1,106 @@
+
+Implement an option to disable required_version checking.
+
+Try to guess the equivalent OpenTofu version for a specified Terraform version.
```
IMO, we should favor being explicit over being implicit.

cam72cam replied:
struck out in b3b094db80 with explanation

ollevche commented on rfc/20240610-Tofu-Version-Compatibility.md:
```diff
@@ -0,0 +1,106 @@
+
+Should there be an option (config, CLI, or ENV) to:
+* Disable the version checking altogether?
+* Don't check terraform required_versions
```
This option would just ignore the warning, right?

cam72cam replied:
Correct

siddharthasonker95 commented on rfc/20240610-Tofu-Version-Compatibility.md:
```diff
@@ -0,0 +1,106 @@
+* Disable the version checking altogether?
+* Don't check terraform required_versions
+* Treat both Tofu and Terraform version as identical instead of using our equivalent "terraform version" guess?
+* Do we show the guessed terraform version in `tofu -version`, maybe with a `-verbose` flag?
```
This can be a good add on, as displaying the guessed Terraform version provides users with immediate feedback and ensures they are using the correct version for their configurations.

cam72cam replied:
@janosdebugs above reminded us that the TSC does not want a version compatibility table.   Perhaps we could link to the migration docs instead?

RLRabinowitz commented on rfc/20240610-Tofu-Version-Compatibility.md:
```diff
@@ -0,0 +1,106 @@
+> [!NOTE]
+> As we are unable to inspect the Hashicorp Terraform Repository in depth, all version compatibility must be based on Release Notes / Changelog Entries as well as user reports.
+
+## Proposed Solution
```
Maybe I missed where it is said in this RFC, but I couldn't find any place mentioning that the approach is to ignore the version constraint altogether or not. It's also a bit unclear whether this is the approach in this RFC, or if it is to maintain a compatibility table with Terraform versions

cam72cam replied:
Hopefully clarified in b3b094db80

cam72cam commented on rfc/20240610-Tofu-Version-Compatibility.md:
```diff
@@ -0,0 +1,106 @@
+
+When loaded by:
+* Terraform, the `version.tofu` file would be ignored and the "Terraform Required Version" would be 1.7.4.
+* OpenTofu, the `version.tofu` file would be loaded instead of `version.tf` and the "OpenTofu Required Version" would be 1.6.5.
```
Is there any reason that we would not want to produce the warning below when someone uses `terraform -> required_version` even when `tofu -> required_version` has been specified?

cam72cam replied:
Rewritten

cam72cam committed:
Updates per RFC review process

Signed-off-by: Christian Mesh <christianmesh1@gmail.com>

cam72cam commented on rfc/20240610-Tofu-Version-Compatibility.md:
```diff
@@ -0,0 +1,106 @@
+Should there be an option (config, CLI, or ENV) to:
+* Disable the version checking altogether?
+* Don't check terraform required_versions
+* Treat both Tofu and Terraform version as identical instead of using our equivalent "terraform version" guess?
```
I struck out this option and added some reasoning in b3b094db80

cam72cam commented on rfc/20240610-Tofu-Version-Compatibility.md:
```diff
@@ -0,0 +1,106 @@
+
+The `configs.Module.CheckCoreVersionRequirements()` function is where we will need to emit the warnings and corresponding debug log entries.
+
+### Open Questions
```
Good catch! b3b094db80

ImIOImI commented on rfc/20240610-Tofu-Version-Compatibility.md:
```diff
@@ -0,0 +1,109 @@
+
+The `configs.Module.CheckCoreVersionRequirements()` function is where we will need to emit the warnings and corresponding debug log entries.
+
+As discussed in the `.tofu` RFC, we will want to quickly work on support for this featu
```
Spelling... featu => feature

cam72cam committed:
Review feedback

Signed-off-by: Christian Mesh <christianmesh1@gmail.com>

cam72cam committed:
Revise Tofu Version Compatibility RFC

As we have received more feedback from the community and now have a
better idea of the pace of adoption of the .tofu extension, a clearer
view of what users actually need to work around version constraints is
available.

Signed-off-by: Christian Mesh <christianmesh1@gmail.com>

apparentlymart commented:
I found this older proposal just due to the rebase today, so I'm sorry that I'm coming in late and so might  be covering ground already covered, but I have a "Potential Alternative" that seems worth discussing in the relevant section even if we don't decide to do it:

The following is starting from the idea that for any version number v1.6 or greater there is (more-or-less) no well-defined relationship between Terraform version numbers and OpenTofu version numbers.

Given that, I'd consider trying for a rule that achieves some behavior like the following, for situations where `required_version` is only declared in a `.tf` file and there is no overriding `.tofu` file:

- If the `required_version` only uses version bounds in 1.5.x or earlier, treat it as an OpenTofu version constraint.
- If the `required_version` mentions any version bound 1.6.x or later, assume it's for Terraform and have OpenTofu entirely ignore it.

The above rules are naive and simplistic, and so would probably require some further nuance in practice.

Some common situations that I think _would_ be handled by these rules are:

- Exact version: `1.5.0` (though of course this would always fail in OpenTofu, since there _is_ no OpenTofu 1.5)
- Lower-bound only, acceptable: `>= 1.5.0` (if we assume that OpenTofu remains backward-compatible with Terraform 1.5 and earlier -- which is our goal -- then it's fine for OpenTofu to accept this version constraint as matching all available OpenTofu releases)
- Lower-bound only, ignored: `>= 1.6.0` (assume that this is talking about _Terraform_ 1.6.0, because it's in a `.tf` file, and so this module has effectively no constraint at all on which OpenTofu versions are compatible)
- Pessimistic: `~> 1.5.0` (again, this would always fail in OpenTofu because there are no OpenTofu versions that match this constraint)

The trickier cases are constraints like `>= 1.5.0, < 1.7.0`. What is the author intending to communicate with a version constraint like that?

---

Overall I feel like the risk of ignoring a version constraint we're unsure about might be a little overstated in this proposal. Version constraints like this are -- in public shared modules, at least -- most commonly used to give better feedback when a module is using a feature that would cause some sort of syntax or semantic error in older versions, since "you are using the wrong version" is a clearer signal than a random syntax error.

It is true that sometimes people use `terraform_version` in _root modules_ with an exact constraint or pessimistic constraint to prevent mistakes where operators might use the wrong version. In that situation there _is_ some risk that someone might accidentally use OpenTofu instead of Terraform or vice-versa, but there's only so much we can do about that as long as Terraform itself won't reject a module that's written for OpenTofu.

Given that, I'd personally hope for a solution more like my alternative above where we just make a best effort to interpret what's communicated in a module but "fail open" if a module doesn't seem to have been written with OpenTofu in mind, rather than introducing obscure language features that allow one module to change how another module's version constraints are interpreted.

All of this is opinions weakly held, though. I don't have any major objections to the proposal, but if we've considered a more "best effort" solution like I described above and have rejected it then I'd like to see some words about _why_ we rejected it in the "Potential Alternatives" section. It might also help to be more explicit about what specific risks we are imagining when asserting that this is "potentially dangerous option", since not all dangers are equal in weight.

cam72cam commented:
@apparentlymart you are hitting this RFC at just the right time, no need to apologize.  I ended up rewriting the RFC this morning based on the original conversation and lessons learned since.

I didn't explicitly say this in the proposal, but my potentially naive approach would be to only flag "Terraform Versions" if the required_version check failed.  I think that in practice, it would behave similarly to what you have described above.

I see it more as a flag that says "I know that there's a `Terraform Version` that's invalid within my project / module and I'm choosing to tell opentofu that it's not".  It would also not impact the root module's direct version constraint check (I think).

This kind of gets to my thoughts toward the end on of this should just be enabled by default.  As there is no well defined relationship between terraform and opentofu versions, the "Terraform Version" in the "OpenTofu" project doesn't really have any meaning.

My hesitation in just enabling this unconditionally is that some people may? be relying on some perceived feature compatibility in past terraform and opentofu versions.

cam72cam committed:
Add Open Question from RFC PR discussion

Signed-off-by: Christian Mesh <christianmesh1@gmail.com>

cam72cam commented:
@abstractionfactory I've got a local checkout I can dig though and post a report here.

cam72cam commented:
Here are the stats:
> Total: 33119, preRelease: 15309, preFork: 30493, postFork: 31182, unknown: 6
> preReleaseOnly: 1330, preForkOnly: 593, postForkOnly: 1290

Here is a list of all of the modules that have upgraded to post fork requirements:
https://gist.github.com/cam72cam/c9eac230887c53e69ed13ed2f6e2082a

cam72cam commented:
With that data in mind, it's clear that modules are being updated to newer terraform/tofu versions and we should have something in place to handle this situation.

cam72cam commented:
I don't follow @abstractionfactory?  Why would we exclude version requirements that can still have critical bugfixes?

cam72cam commented:
Part of that depends on our release cycle.  910 out of the 1290 post-fork-only versions require 1.9 and would *not* function prior to last week and would have been out of luck (other than a fork) for about 6 months.

Even still, the "Terraform Version" identified there is still not explicitly representing a "Tofu Version" and bug-for-bug compatibility is not certain.  As time goes on it will mean less and less.

I understand your hesitance about "modifying" modules from the outside, and I have expressed similar concerns in the past.  The question in my mind is how we treat an existing feature that is known to be misleading and potentially dangerous.  That's the root of the open question IMO and even if we don't merge this RFC (in any form) I'd like to have a clear understanding (documented) idea of what "Terraform Version" means in the context of OpenTofu

ollevche commented on rfc/20240610-Tofu-Version-Compatibility.md:
```diff
@@ -0,0 +1,232 @@
+1) Disabled for all child modules of the current module.  This represents someone who has lots of child modules that have not yet added "Tofu Versions" and they are comfortable with the risk of bypassing the "Terraform Version" in all of them (and their children).
+2) Disabled for a specific child module of the current module.  This represents someone who is pulling a small number of child modules that have not yet added "Tofu Versions" and they would like to limit the blast radius of ignoring the "Terraform Version" check in this limited area of the current module.
+
+For for the first scenario, a flag called "bypass_terraform_version_requirement" is added to the `product{}` block.  If set to true, all child modules of the current module will only error on "Tofu Versions" and instead produce warnings for "Terraform Versions".
```
```suggestion
For the first scenario, a flag called "bypass_terraform_version_requirement" is added to the `product{}` block.  If set to true, all child modules of the current module will only error on "Tofu Versions" and instead produce warnings for "Terraform Versions".
```

ollevche commented:
I would like to summarize the ideas I just read and propose a combined version of what's been discussed already.

Given that we don't have a mapping of tf <> tofu versions higher than the forked one, I agree that we should somehow distinguish between what variant the original author meant exactly (tofu or terraform). This is currently possible via `.tofu` files extension, however, it requires module author to alter the configuration.

I also agree that we shouldn't allow altering module behavior outside of the module itself (i.e. via `bypass_terraform_version_requirement`). I support the idea that after-fork version constraints are irrelevant for OpenTofu if there is no clear sign that author meant tofu variant exactly. That's being said, it seems reasonable to me to start deprecating `terraform` block and introduce a `tofu` one.

The first step could be introducing a CLI flag to ignore the `terraform` block version constraint (in the root and all the child modules) and rely on the one specified in `tofu` block. The check could still provide a warning if there is only `terraform` block version constraint (basically, opt-in version of what Martin described).

I think this is not as risky as just starting to ignore the checks, because user needs to opt-in. Also, it doesn't require module authors to change their configuration immediately.

Once we see wide adoption of `tofu` block, we could completely deprecate `terraform` one to minimize the confusion.

WDYT?

cam72cam committed:
Update rfc/20240610-Tofu-Version-Compatibility.md

Co-authored-by: Oleksandr Levchenkov <ollevche@gmail.com>
Signed-off-by: Christian Mesh <christianmesh1@gmail.com>

cam72cam commented:
I think that makes a lot of sense @ollevche and matches my understanding of the issue.  The question would then become what would the opt-in look like?  Off the top of my head we have a flag in the `tofu{}` block, a .tofurc flag, an environment variable, and a cli flag.

I'm kind of partial to the `tofu{}` block solution if we were to go this route.  It would clearly opt a project into the more visible deprecation (warnings for now) treatment of the `terraform {}` block

ollevche commented:
I think specifying `tofu` block should be treated as opt-in (on per module basis), however it is not going to resolve the issue with the child modules with only `terraform` block. For that case, it shouldn't be in the configuration IMO. CLI flag, env var or .tofurc flag are nice options.

ImIOImI commented:
Personally speaking, I understand the utility of using a `versions.tofu` file to override version requirements... but I can say from personal experience that overrides in configs very often confuse people. I've laid out some rather simple override patterns to see them become completely confused and ignored by colleagues multiple times.

If we can have all the version requirements in a `.tf` file living side by side with the rest of the tofu code, I think that would be best from a code readability standpoint. IMO, I think that would be a much more explicit signal about what's actually happening.

ollevche commented on rfc/20240610-Tofu-Version-Compatibility.md:
```diff
@@ -0,0 +1,232 @@
+1) Disabled for all child modules of the current module.  This represents someone who has lots of child modules that have not yet added "Tofu Versions" and they are comfortable with the risk of bypassing the "Terraform Version" in all of them (and their children).
+2) Disabled for a specific child module of the current module.  This represents someone who is pulling a small number of child modules that have not yet added "Tofu Versions" and they would like to limit the blast radius of ignoring the "Terraform Version" check in this limited area of the current module.
+
+For the first scenario, a flag called "bypass_terraform_version_requirement" is added to the `product{}` block.  If set to true, all child modules of the current module will only error on "Tofu Versions" and instead produce warnings for "Terraform Versions".
```
```suggestion
For the first scenario, a flag called "bypass_terraform_version_requirement" is added to the `project{}` block.  If set to true, all child modules of the current module will only error on "Tofu Versions" and instead produce warnings for "Terraform Versions".
```

Gogotchuri commented on rfc/20240610-Tofu-Version-Compatibility.md:
````diff
@@ -0,0 +1,232 @@
+```
+
+
+#### Product scoped behavior:
````
```suggestion
#### Module-scoped behavior:
```

cam72cam commented:
Closing this as subsumed by https://github.com/opentofu/opentofu/pull/3085
